### ISSUE TITLE
[Feature Request] Use FP16 Precision for VR180 Pipeline to Reduce VRAM Usage

### ISSUE THREAD

**Issue by Umstead** [VisoMasterFusion/VisoMaster-Fusion#243]
## Summary

Replace `float32` with `float16` (FP16) in the VR180 equirectangular projection pipeline (`Equirec2Perspec_vr.py`, `Perspec2Equirec_vr.py`, `vr_utils.py`) to halve VRAM usage for intermediate floating-point tensors without affecting visual quality. This directly reduces peak VRAM by ~1.4 GB during VR180 processing on 6K/8K content.

---

## Motivation

Processing 6K/8K VR180 video requires significant GPU VRAM. While other parts of the pipeline (RetinaFace detection, TensorRT/ONNX inference) already benefit from FP16 or are inference-engine-constrained, the VR projection math (equirectangular-to-perspective sampling, perspective-to-equirectangular blending, and feathered stitch masks) still operates entirely in float32. This is unnecessary overhead — the math involved is well within FP16 range.

### Current State

The VR180 pipeline processes equirectangular (EQR) frames through these steps:

1. **Equirect2Perspec (E2P)**: Extracts per-eye perspective crops from the EQR frame using `grid_sample`. The cached float32 image buffer (`_img_float`) and the coordinate grid are both float32.
2. **Perspec2Equirec (P2E)**: Merges processed perspective crops back into the EQR frame using `grid_sample`. The cached float32 image buffer and coordinate grid are both float32.
3. **Stitch Blending**: Blends perspective crops at tile boundaries using feathered alpha masks. Both the target frame and component frame are converted to float32.

### Why FP16 is Safe Here

| Concern | Analysis | Verdict |
|---------|----------|---------|
| **grid_sample bilinear interpolation** | FP16 quantization error < 1/256 ≈ 0.4%, well below uint8 rounding threshold |  Safe |
| **Coordinate grids (in [-1, 1])** | FP16 exact range is ±1024, so ±1.0 is perfectly represented |  Safe |
| **Stitch blend math** | Max intermediate value is 255 (uint8 range × mask ≤ 1.0). FP16 can exactly represent all integers from 0 to 2048 |  Safe |
| **Cumulative error over frames** | Each frame uses freshly converted FP16 data; no cross-frame accumulation |  Safe |
| **Comparison with RetinaFace** | RetinaFace is already running in FP16 (`FP16 Acceleration ENABLED for RetinaFace`) |  Validated precedent |

### Quantified VRAM Savings

| Location | Tensor | Resolution | Float32 | Float16 | Savings |
|----------|--------|-----------|--------|--------|---------|
| E2P `_img_float` cache | Per-frame EQR | 6144×3072 | ~190 MB | ~95 MB | **~95 MB** |
| P2E `_img_tensor_cxhxw_rgb_float` | Per-frame EQR | 6144×3072 | ~190 MB | ~95 MB | **~95 MB** |
| P2E `equirect_component_float` | Per-tile sampling output | varies | ~190 MB | ~95 MB | **~95 MB** |
| Stitch `target_float` + `component_float` | Blend intermediates | 6144×3072 | ~380 MB | ~190 MB | **~190 MB** |
| **Total per-frame peak** | | | **~950 MB** | **~475 MB** | **~475 MB** |

> **Note**: These are peak intermediate allocations. The actual VRAM reduction observed is ~1.3–1.4 GB on 6K VR180 content due to overlapping reductions across workers.

### Side Benefits

1. **Faster processing**: FP16 tensor ops on NVIDIA GPUs (especially Tensor Cores) are 2–8× faster than FP32.
2. **Reduced PCIe traffic**: Smaller tensors mean less data movement between CPU and GPU.
3. **No accuracy loss**: The 0.4% quantization error from grid_sample is imperceptible in uint8 output.

---

### 1. `app/processors/external/Equirec2Perspec_vr.py`

Change the per-frame float cache from float32 to float16:

```python
# Before:
self._img_float = self._img_tensor_cxhxw_rgb_uint8.float() * (1.0 / 255.0)

# After:
self._img_float = self._img_tensor_cxhxw_rgb_uint8.half() * (1.0 / 255.0)
```

Change the coordinate grid dtype to match input:

```python
# Before:
grid = torch.stack((grid_x, grid_y), dim=2).unsqueeze(0)  # float32

# After:
grid = torch.stack((grid_x, grid_y), dim=2).unsqueeze(0).half()  # float16, matches input
```

### 2. `app/processors/external/Perspec2Equirec_vr.py`

Change the per-frame float buffer from float32 to float16:

```python
# Before:
self._img_tensor_cxhxw_rgb_float = img_tensor_cxhxw_rgb_uint8.float() / 255.0

# After:
self._img_tensor_cxhxw_rgb_float = img_tensor_cxhxw_rgb_uint8.half() / 255.0
```

Change the coordinate grid dtype to match input:

```python
# Before:
grid = torch.stack((grid_x_persp, grid_y_persp), dim=2).unsqueeze(0)  # float32

# After:
grid = torch.stack((grid_x_persp, grid_y_persp), dim=2).unsqueeze(0).half()  # float16
```

### 3. `app/helpers/vr_utils.py`

Change stitch blending intermediates from float32 to float16:

```python
# Before:
target_float = target_equirect_torch_cxhxw_rgb_uint8.float()
component_float = equirect_component_torch.float()

# After:
target_float = target_equirect_torch_cxhxw_rgb_uint8.half()
component_float = equirect_component_torch.half()
```

### 4. Optional: Additional VR Cache Reductions (complementary)

These cache size reductions complement the FP16 change:

```python
# Perspec2Equirec_vr.py
_P2E_GRID_MASK_CACHE_MAX = 4  # → reduce to 2 or 1

# vr_utils.py
_FEATHERED_MASK_CACHE_MAX = 8  # → reduce to 4 or 2
```

---

## Test Results

| Configuration | VRAM (Before) | VRAM (After) | Savings |
|---------------|---------------|--------------|---------|
| VR180 6K, Both Eyes (2 workers) | ~15.85 GB | ~14.5 GB | ~1.3 GB |
| VR180 6K, Low VRAM - Both Eyes (1 worker) | ~14.5 GB | ~12.7 GB | ~1.8 GB |

Processing speed remained stable with no quality degradation observed.

---

## Backward Compatibility

- The change is **transparent** — no new settings, no API changes.
- The change is **safe** — FP16 is sufficient precision for all operations in the VR pipeline.
- If a future GPU doesn't support FP16, PyTorch will raise an error at the `.half()` call, which is preferable to silent precision loss.

---

## Priority

**Medium**. The optimization is straightforward and well-isolated, but the VR pipeline is only a fraction of total VRAM usage (the deepfake model itself dominates). The primary value is enabling 6K/8K VR180 processing on 12 GB GPUs that previously required Low VRAM mode.

---

## Tags

`vr180` `vram` `fp16` `performance` `optimization`

---

## References

- PyTorch `torch.half()` / `.half()` documentation
- PyTorch `torch.grid_sample()` — accepts both float32 and float16 inputs
- NVIDIA Tensor Core documentation — FP16 inference optimizations

**Comment by Elricfae** [comment]
GlatOs will surely work on improving the VR pipeline, we also would like to include part of the pipeline to the sequential detector to gain the same improvement it gives.
Thanks for letting us now about the FP16 reduction.

**Comment by Umstead** [comment]
> GlatOs will surely work on improving the VR pipeline, we also would like to include part of the pipeline to the sequential detector to gain the same improvement it gives. Thanks for letting us now about the FP16 reduction.


Thanks for the update! Great to hear you plan to improve the VR pipeline and integrate it into the sequential detector.

I’m glad my report on FP16 reduction was helpful. Looking forward to the optimizations!
